src/Core/Checkout/Document/Zugferd/ZugferdBuilder.php

```diff
@@ -0,0 +1,82 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Checkout\Document\Zugferd;
+
+use horstoeko\zugferd\ZugferdDocumentBuilder;
+use horstoeko\zugferd\ZugferdProfiles;
+use Shopware\Core\Checkout\Cart\LineItem\LineItem;
+use Shopware\Core\Checkout\Cart\Price\Struct\CartPrice;
+use Shopware\Core\Checkout\Document\DocumentConfiguration;
+use Shopware\Core\Checkout\Document\Struct\DocumentGenerateOperation;
+use Shopware\Core\Checkout\Order\Aggregate\OrderAddress\OrderAddressEntity;
+use Shopware\Core\Checkout\Order\Aggregate\OrderCustomer\OrderCustomerEntity;
+use Shopware\Core\Checkout\Order\Aggregate\OrderDelivery\OrderDeliveryCollection;
+use Shopware\Core\Checkout\Order\Aggregate\OrderLineItem\OrderLineItemCollection;
+use Shopware\Core\Checkout\Order\Aggregate\OrderLineItem\OrderLineItemEntity;
+use Shopware\Core\Checkout\Order\OrderEntity;
+use Shopware\Core\Framework\Context;
+use Shopware\Core\Framework\Log\Package;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+#[Package('after-sales')]
+class ZugferdBuilder
+{
+    /**
+     * @internal
+     */
+    public function __construct(
+        protected EventDispatcherInterface $eventDispatcher
+    ) {
+    }
+
+    public function buildDocument(OrderEntity $order, DocumentGenerateOperation $operation, DocumentConfiguration $config, Context $context): string
```

Comment:
I think we discussed this in Slack but just so it's public. I think this can be done differently as such:

```suggestion
    public function buildDocument(RenderedDocument $document): string
```

More [here](https://github.com/shopware/shopware/blob/e7a9517a84cceb56dc9502f44f428c973e936a1c/src/Core/Checkout/Document/Zugferd/ZugferdBuilder.php#L30).

Obviously it will need some minor changes a step earlier.

Is there a reason you don't want to do it this way? Or just an oversight?

Reply:
Firstly overseen. But then I saw, that I could remove the `DocumentGenerateOperation` parameter.
With this we have `OrderEntity`, `DocumentConfiguration` and the usual `Context`. In my opinion, this makes it clearer what you need when you want to call the method.

Another point. We have the `documentConfig` object (with `documentConfigLoader->load()`), convert it into an array, to convert it back into the object.

Writing the test for the `RenderedDocument` 
```php
$rendered = new RenderedDocument(
    config: $documentConfig->jsonSerialize()
);
$rendered->setOrder($order);
$rendered->setContext(Context::createDefaultContext());

buildDocument($rendered)
// instead of
buildDocument($order, $documentConfig, Context::createDefaultContext())
```

Reply:
I don't view how a test is written as a valid argument for how code is structured. 95% of tests won't be written the same way the code using the tested method is being written.

My feeling around this is that the `Order` and `Context` is available on the `RenderedDocument`, which means you are passing the same object around and just getting the values off that.

If you are fine with the current implementation, then I'm fine with it too.